scripts/tuple_tree_generator/tuple_tree_generator/templates/tuple_tree_gen.prelude.ts

```diff
@@ -0,0 +1,512 @@
+}
+
+export function genGuid(): bigint {
+    return randomBigint(2n ** 10n + 1n, 2n ** 64n - 1n);
+}
+
+// Tree traversal
+
+function getElementByPathArray<T>(path: string[], obj: any): T | undefined {
+  const component = path[0];
+  if (obj instanceof Array) {
+    for (const elem of obj) {
+      if (
+        "key" in (elem as any) &&
+        typeof elem["key"] == "function" &&
+        elem["key"]() == component
+      ) {
+        if (path.length == 1) {
+          return elem as T;
+        } else {
+          return getElementByPathArray(path.slice(1), elem);
+        }
+      }
+    }
+    return undefined;
+  } else {
+    if (component in obj) {
+      if (path.length == 1) {
+        return obj[component];
+      } else {
+        return getElementByPathArray(path.slice(1), obj[component]);
+      }
+    } else {
+      return undefined;
+    }
+  }
+}
+
+export function getElementByPath<T>(path: string, obj: any): T {
+  return getElementByPathArray(path.split("/").slice(1), obj);
+}
+
+// Tree Diffing
+
+interface IDiffSet {
+  Changes: IDiff[];
+}
+
+export class DiffSet {
+  Changes: Diff[];
+
+  constructor(Changes: Diff[]) {
+    this.Changes = Changes;
+  }
+
+  static parse(rawObject: IDiffSet): DiffSet {
+    return new DiffSet(
+      rawObject.Changes.map((e) => new Diff(e.Path, e.Add, e.Remove))
+    );
+  }
+
+  toJSON(): IDiffSet {
+    return {
+      Changes: this.Changes.filter((e) => e.isValid()).map((e) => e.toJSON()),
+    };
+  }
+
+  reversed(): DiffSet {
+    return new DiffSet(
+      this.Changes.map((e) => new Diff(e.Path, e.Remove, e.Add))
+    );
+  }
+
+  static deserialize(input: string): DiffSet {
+    return DiffSet.parse(yaml.parse(input));
+  }
+
+  serialize(): string {
+    const parseOptions = {
+      keepUndefined: false,
+      sortMapEntries: false,
+    } as const;
+
+    const options = {
+      doubleQuotedAsJSON: false,
+      defaultKeyType: yaml.Scalar.PLAIN,
+      defaultStringType: yaml.Scalar.QUOTE_DOUBLE,
+      directives: true,
+    } as const;
+
+    const doc = new yaml.Document(this, parseOptions);
+    return doc.toString(options);
+  }
+}
+
+interface IDiff {
+  Path: string;
+  Add?: any;
+  Remove?: any;
+}
+
+class Diff {
+  Path: string;
+  Add: any;
+  Remove: any;
+
+  constructor(Path: string, Add: any, Remove: any) {
+    this.Path = Path;
+    this.Add = Add;
+    this.Remove = Remove;
+  }
+
+  toJSON(): IDiff {
+    let ret = { Path: this.Path };
+    if (this.Add !== undefined) {
+      ret["Add"] = this.Add;
+    }
+    if (this.Remove !== undefined) {
+      ret["Remove"] = this.Remove;
+    }
+    return ret;
+  }
+
+  isValid(): boolean {
+    return this.Add !== undefined || this.Remove !== undefined;
+  }
+}
+
+interface Comm<T> {
+  only1: Set<T>;
+  only2: Set<T>;
+  common: Set<T>;
+}
+
+function isNull(obj: any): obj is undefined | null {
+  return obj === undefined || obj === null;
+}
+
+function objComm(obj1: any, obj2: any): Comm<string> {
```

Comment:
`Comm`? What's that?

Reply:
https://man7.org/linux/man-pages/man1/comm.1.html
It's a linux tool that given two files it splits it in 3 columns (only 1) (only 2) (common)
The behavior here is similar where I need to know keys that are only in `obj1`, `obj2` and in common

Reply:
Uhm, OK, add documentation about that. It's not intuitive.

Reply:
Added docstrings